### PR TITLE
Add `RuntimeEnv::try_new` and deprecate `RuntimeEnv::new`

### LINKED ISSUE #12554: Make it clearer that `RuntimeEnv::new()` is fallable
### Is your feature request related to a problem or challenge?

While working on https://github.com/datafusion-contrib/datafusion-dft it was not clear that [`RuntimeEnv::new()`](https://docs.rs/datafusion/latest/datafusion/execution/runtime_env/struct.RuntimeEnv.html#method.new) actually returns a `Result` as the typical rust convention is a function `new()` is infallable and a function like `try_new()` is fallable. While this isn't a big deal it was a small usability papercut

### Describe the solution you'd like

I would like `RuntimeEnv` to follow standard rust conventions to make it easier to use

### Describe alternatives you've considered

I suggest:
1. Deprecating [`RuntimeEnv::new()`](https://docs.rs/datafusion/latest/datafusion/execution/runtime_env/struct.RuntimeEnv.html#method.new)
2. Add a new function `RuntimeEnv::try_new()` that does the same thing
3. Update all the datafusion code to use `try_new` instead

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by OussamaSaoudi**
## Which issue does this PR close?

Closes #12554

## What changes are included in this PR?

- Move existing functionality of `RuntimeEnv::new` to `RuntimeEnv::try_new`
- Replace all calls to `RuntimeEnv::new` in code and documentation to use `RuntimeEnv::try_new`

## Are these changes tested?

No extra tests required. Ensured that compilation, existing tests, and doc tests pass.
## Are there any user-facing changes?
None.

### CODE REVIEW

File: datafusion/execution/src/runtime_env.rs [hunk 1]
```diff
@@ -66,8 +66,12 @@ impl Debug for RuntimeEnv {
 }
 
 impl RuntimeEnv {
-    /// Create env based on configuration
+    #[deprecated(note = "please use `try_new` instead")]
```
Comment (alamb) [inline datafusion/execution/src/runtime_env.rs:69]: ❤️

**Review by alamb — COMMENTED** [review]
Thank you @OussamaSaoudi -- I have kicked off the CI tests for this one

**Review by alamb — APPROVED** [review]
👨‍🍳 👌 

Very nice @OussamaSaoudi  -- looks great to me. 

I am surprised this didn't cause any other changes

**Comment by alamb** [comment]
Thanks again @OussamaSaoudi

**Comment by OussamaSaoudi** [comment]
Thanks for the clear and beginner friendly issues @alamb! This was a great way to get introduced to the code, and I'm looking forward to contributing more :D

**Comment by alamb** [comment]
> Thanks for the clear and beginner friendly issues @alamb! This was a great way to get introduced to the code, and I'm looking forward to contributing more :D

You are welcome -- and I am glad the tickets fulfilled their intended purpose

**Pull request merged**